lib/Doctrine/ORM/EntityManager.php

```diff
@@ -157,8 +156,12 @@ protected function __construct(Connection $conn, Configuration $config, EventMan
         $this->metadataFactory->setCacheDriver($this->config->getMetadataCacheImpl());
 
         $this->repositoryFactory = $config->getRepositoryFactory();
-        $this->unitOfWork        = new UnitOfWork($this);
-        $this->proxyFactory      = new ProxyFactory(
+
+        $unitOfWorkClassName  = $config->getUnitOfWorkClassName();
+        $this->unitOfWork     = new $unitOfWorkClassName($this);
```

Comment:
@stof the reason behind configuring UnitOfWork class name is that next layer (UnitOfWOrk) depends on higher level (EntityManager). So refactoring of these 2 layers would be a really major one.